### PR TITLE
KIC-KGO integration

### LINKED ISSUE #2913: KIC's Gateway API able to work under KGO
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Problem Statement

This issue is needed to make the kong-gateway-operator (KGO) work with `HTTPRoutes`: as a KGO user I want to deploy a Gateway, and then once it's ready I want to deploy HTTPRoutes which will be served by that Gateway. To achieve so, we need to remove the KIC requirement that Gateway resources be unmanaged.

### Proposed Solution

_No response_

### Additional information

related KGO issue: https://github.com/Kong/gateway-operator/issues/188

### Acceptance Criteria

- [ ] We should allow the injection of the controller-name via args
- [ ] The gateway resources shouldn't be unmanaged to be successfully reconciled

### PR DESCRIPTION
**Pull request by mlavacca**
**What this PR does / why we need it**:
Fixes #2913 

**Which issue this PR fixes**:

**Special notes for your reviewer**:

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [x] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

### CODE REVIEW

File: internal/manager/run.go
Comment (shaneutt) [inline internal/manager/run.go:243]: You should use a permalink here: as `main` is subject to significant change.

File: internal/manager/run.go
Comment (shaneutt) [inline internal/manager/run.go:246]: This whole function _might_ be reasonable to add as an upstream utility if you felt so inclined. Upstream has recently started allowing "utility functions" for the Go lib.

File: test/e2e/features_test.go
Comment (shaneutt) [inline test/e2e/features_test.go:613]: I'm a little confused by this test, it's not _actually_ testing a **managed** gateway right?

I just realized something else: the HTTPRoute should not be able to attach without the `GatewayClass` status becoming ready for use... do we have contention between the KGO and the KIC where _both_ are trying to manage the status of the `GatewayClass`? We may need to add the `unmanaged` annotation to the `GatewayClass` as a means to not have both controllers colliding on those.

File: internal/manager/run.go
Comment (mlavacca) [reply internal/manager/run.go:246]: Sure, I'll open an issue upstream about this. Can we keep this one in the meantime?

File: test/e2e/features_test.go
Comment (mlavacca) [reply test/e2e/features_test.go:613]: As far as I can see in KGO we never update the GatewayClass status, there shouldn't be any collision

Comment (mlavacca) [reply test/e2e/features_test.go:613]: I'm not convinced too about this test. I want to test managed gateways (I changed the helper), but it is just an attempt; it is not ready yet. I'm afraid that without an actor reconciling the Gateway status, the HTTPRoute could never get ready. So probably, we should perform an e2e test in KGO to test this functionality for real.

Comment (mlavacca) [reply test/e2e/features_test.go]: I dropped the test

File: internal/manager/run.go
Comment (shaneutt) [reply internal/manager/run.go]: Yep

**Review by shaneutt — CHANGES_REQUESTED** [review]
So what I'm starting to realize is that we have a couple problems here.

Firstly, take a look at https://github.com/Kong/kubernetes-ingress-controller/blob/15460a3fcc906c9b07cd79985bbdbe5cc01fa778/internal/controllers/gateway/gateway_controller.go#L342. What we're doing here is if a `Gateway` is configured with our `controllername` we just _assume_ it's unmanaged mode, and add the annotations ourselves. I'm not exactly sure when or how this behavior came to be but it wasn't the original intent. As a part of this PR, we need to have the gateway controller _not reconcile any Gateway that doesn't already have the annotation_ (though we should still support the `true` value).

Also related to our previous conversation about the `GatewayClass`, we are indeed updating the status of the `GatewayClass` if it has our `controllername`:

https://github.com/Kong/kubernetes-ingress-controller/blob/15460a3fcc906c9b07cd79985bbdbe5cc01fa778/internal/controllers/gateway/gatewayclass_controller.go#L93

This will end up having some conflicts with the KGO [which needs to update this status](https://github.com/Kong/gateway-operator/issues/273). We need to find an approach where we ignore the `GatewayClass` unless it's specifically being used for an unmanaged mode `Gateway`. The most straightforward way I can think to do this would be to require an annotation like unmanaged mode on the gatewayclass as well as the Gateways attached to it. This is something of a breaking change, which fortunately we're catching while we're still in the alpha stage. The implications are that we would have to update a multitude of tests and documentation. Alternatively, we could consider the `GatewayClass` only accepted once an `unmanaged` mode `Gateway` becomes attached to it, however this is a bit clunky (though unmanaged mode in general is a bit clunky so I don't think it's out of the question).

Let me know your thoughts?

File: test/e2e/features_test.go
Comment (shaneutt) [reply test/e2e/features_test.go]: I'll consider this comment resolved, some relevant follow up added in https://github.com/Kong/kubernetes-ingress-controller/pull/2917#pullrequestreview-1102604851

File: internal/manager/config.go
Comment (rainest) [inline internal/manager/config.go:156]: Since we have an effective default value under the hood, we should list it here:

```suggestion
	flagSet.StringVar(&c.GatewayAPIControllerName, "gateway-api-controller-name", "konghq.com/kic-gateway-controller", "The controller name to match on Gateway API resources.")
```

File: internal/manager/run.go
Comment (rainest) [inline internal/manager/run.go:199]: The outer if block here would be unnecessary if we proceed with making the default controller string a CLI default instead of just the initial `gateway.ControllerName` value.

**Review by rainest — CHANGES_REQUESTED** [review]
I don't think there's any reason we wouldn't be able to make the default a default in the CLI, so IMO we should do so for documentation (granted, in practice I don't think anyone's invoking the binary with `--help`, but it's consistent with how we handle ingress class, and maybe we'll find a way to autogenerate https://docs.konghq.com/kubernetes-ingress-controller/2.5.x/references/cli-arguments/ from the code someday).

**Comment by rainest** [comment]
> The most straightforward way I can think to do this would be to require an annotation like unmanaged mode on the gatewayclass as well as the Gateways attached to it.

@shaneutt if we look for it on GatewayClass do we still need to look for it on Gateway also? It seems like we'd be able to choose whether to run the unmanaged handling by getting the GatewayClass (which we do already), checking if it has the annotation, and then treating all Gateways with that class as unmanaged. The breakage is the same (GatewayClass didn't require that before, now it does), but it's maybe less configuration required going forward? The benefits of needing to only configure it on the GatewayClass once are hampered by unmanaged not supporting more than one Gateway anyway, but at least you don't need to remember to configure the annotation on both.

**Comment by shaneutt** [comment]
> > The most straightforward way I can think to do this would be to require an annotation like unmanaged mode on the gatewayclass as well as the Gateways attached to it.
> 
> @shaneutt if we look for it on GatewayClass do we still need to look for it on Gateway also? It seems like we'd be able to choose whether to run the unmanaged handling by getting the GatewayClass (which we do already), checking if it has the annotation, and then treating all Gateways with that class as unmanaged. The breakage is the same (GatewayClass didn't require that before, now it does), but it's maybe less configuration required going forward? The benefits of needing to only configure it on the GatewayClass once are hampered by unmanaged not supporting more than one Gateway anyway, but at least you don't need to remember to configure the annotation on both.

Going forward the annotation _only_ has the effect of triggering (or not triggering) reconciliation for `Gateway` resources, and there are no implications on attaching an `*Route` to a `Gateway` that was created elsewhere now. As such I would say it should be viable to move the annotation to the GWC as you suggest after considering this and looking over the codebase I can't see any fallout from this.

**Comment by mlavacca** [comment]
@shaneutt @rainest, we could overturn the problem, and instead of relying on an `unmanaged` annotation in KIC to reconcile the GWC, we can depend on the nonexistence of a `managed` label to do the same thing. KGO would be in charge of attaching this annotation to the GWC, and KIC just remains dormant for GWC with such an annotation. The Gateway `unmanaged` approach isn't touched, and we move the breaking part to KGO, where we don't have the problem of breaking changes yet.

**Comment by shaneutt** [comment]
> @shaneutt @rainest, we could overturn the problem, and instead of relying on an `unmanaged` annotation in KIC to reconcile the GWC, we can depend on the nonexistence of a `managed` label to do the same thing. KGO would be in charge of attaching this annotation to the GWC, and KIC just remains dormant for GWC with such an annotation. The Gateway `unmanaged` approach isn't touched, and we move the breaking part to KGO, where we don't have the problem of breaking changes yet.

Functionally I'm sure this could work, but this "unmanaged" vs "managed" mode is a biproduct of the limitations of the historical KIC deployment mechanisms (e.g. the helm chart and kustomize manifests, and the fact that the KIC wasn't in any position to manage the lifecycle of the Kong Gateway), so ultimately I would like to avoid putting the burden of dealing with this onto the KGO.

**Comment by mlavacca** [comment]
> Functionally I'm sure this could work, but this "unmanaged" vs "managed" mode is a biproduct of the limitations of the historical KIC deployment mechanisms (e.g. the helm chart and kustomize manifests, and the fact that the KIC wasn't in any position to manage the lifecycle of the Kong Gateway), so ultimately I would like to avoid putting the burden of dealing with this onto the KGO.

I get it. Then to summarize, we can move that annotation from the `Gateway` to the `GatewayClasses`. All the non-unmanaged `GatewayClassesese` will not be considered by the respective controller, as well as the `Gateways` using a non-unmanaged `GatewayClass`. To do so, we need to remove the concept of unmanaged `Gateway` and introduce the concept of unmanaged `GatewayClass` and basically change every test of our codebase. Correct?

**Comment by shaneutt** [comment]
> > Functionally I'm sure this could work, but this "unmanaged" vs "managed" mode is a biproduct of the limitations of the historical KIC deployment mechanisms (e.g. the helm chart and kustomize manifests, and the fact that the KIC wasn't in any position to manage the lifecycle of the Kong Gateway), so ultimately I would like to avoid putting the burden of dealing with this onto the KGO.
> 
> I get it. Then to summarize, we can move that annotation from the `Gateway` to the `GatewayClasses`. All the non-unmanaged `GatewayClasses` will not be considered by the respective controller, as well as the `Gateways` using a non-unmanaged `GatewayClass`. To do so, we need to remove the concept of unmanaged `Gateway` and introduce the concept of unmanaged `GatewayClass` and basically change every test of our codebase. Correct?

That sounds right. the `*Route` controllers will _only_ care if they're matching to the right `controllername` via `GatewayClass`, the `GatewayClass` and `Gateway` controllers will now drop any relevant resource if there's no "unmanaged" `GatewayClass` in the mix. :+1: Sorry this ends up being so much more burdensome, but since we're about to release beta it would be good to get this sorted out solidly now rather than having to do it during the beta cycle.

**Comment by mlavacca** [comment]
> That sounds right. the *Route controllers will only care if they're matching to the right controllername via GatewayClass, the GatewayClass and Gateway controllers will now drop any relevant resource if there's no "unmanaged" GatewayClass in the mix. 👍 Sorry this ends up being so much more burdensome, but since we're about to release beta it would be good to get this sorted out solidly now rather than having to do it during the beta cycle.

Ok, clear! I understand that this is important to be addressed now that the Gateway API moves from alpha to beta. I'll update this PR with these changes soon.

**Comment by rainest** [comment]
> instead of relying on an `unmanaged` annotation in KIC to reconcile the GWC, we can depend on the nonexistence of a `managed` label to do the same thing

We should avoid this since the un-managed mode of operation isn't formally defined in the Gateway spec. Managed is effectively the default, assumed state. This is probably a non-issue because it should mostly affect GWCs that we don't interact with at all (because they're for other controllers and we're not configured to use them), but all things being equal otherwise, I'd lean towards marking things explicitly un-managed because it's the unusual case.

File: CHANGELOG.md
Comment (shaneutt) [inline CHANGELOG.md:74]: :+1:

File: internal/controllers/gateway/gateway_controller.go
Comment (shaneutt) [inline internal/controllers/gateway/gateway_controller.go:313]:
```suggestion
	// This check has already been performed by predicates, but we need to ensure this condition
```

**Comment by shaneutt** [comment]
> > Looks like we're on the right path. We updated the examples too, which is important. I'm worried about testing: are there some extra test cases that might be prudent to add? Have we manually tested this? I would like to hear @rainest's thoughts as well.
> 
> Which scenario do you feel we are not testing properly? I think we can add a test case in which a managed GatewayClass is used by the Gateway, and we check that it's not eventually reconciled. WDYT?

Sounds good :+1:

File: internal/controllers/gateway/gateway_controller.go
Comment (randmonkey) [inline internal/controllers/gateway/gateway_controller.go:345]: `GetAnnotations()` is not guaranteed non-nil, so is it possible that `nil` is passed in and cause a panic? If we have some other places to guarantee that `gatewayClass.Annotations` is not `nil`, is it OK to add a commment for it?

**Review by randmonkey — COMMENTED** [review]
Generally OK, but I have a question: `ExtractUnmanagedGatewayClassMode` could cause panic when  `annotations` is `nil`. If we have some mechanism elsewhere to guarantee that `gatewayClass.Annotations` is always non-nil, could we add some comments somewhere to notice that? If not guaranteed, we should judge whether `annotation` is `nil` in that function.

File: CHANGELOG.md
Comment (mlavacca) [reply CHANGELOG.md:74]: I'm using 2917 here

**Review by randmonkey — APPROVED** [review]

**Comment by mlavacca** [comment]
waiting for @shaneutt or @rainest to get the final approval as a double check.

**Review by shaneutt — APPROVED** [review]
I'm approving, but I would like us to wait until we have @rainest's explicit review before we merge as he's spent more time than I in this area of the code recently.

**Review by rainest — APPROVED** [review]
I have? I don't remember anything other than 3.x compatibility issues at this point. I touched status handling stuff but that's only after we've decided to interact with a Gateway.

Semantically we probably should have made separate class/unmanaged checks and called them separately in an umbrella process/don't function, but whatever, that's easy to change later. The tests look appropriate and nothing in the code looks that unusual, :shipit:

**Pull request merged**
